Improve accessibility of the Warning component in the block editor

sarthaknagoshe2002 opened the pull request:
Accessibility Fix: #62737

## What?
This PR improves the Warning component in the block editor by enhancing its accessibility.

## Why?
Currently, the Warning component only provides a visual cue without notifying screen readers, which limits accessibility. These changes ensure that the component is accessible to all users.

## How?

- Added ARIA role="alert" and aria-live="assertive" to the Warning component for screen reader announcements.
- Added aria-describedby="block-editor-warning-message" to the Warning actions.
- Added tabIndex="0", so that the Warning can be accessed using tabs. This helps to navigate inside the warning.

## Testing Instructions

1. Add a More block to a post and copy it using the keyboard shortcut or the Options menu.
2. Paste the copied block elsewhere in the post.
3. Observe the Warning message appears with improved design and triggers a screen reader announcement.

<!-- ## Screenshots or screencast  if applicable -->

sarthaknagoshe2002 committed:
Enhance accessibility of the Warning component

afercia commented:
@sarthaknagoshe2002 thanks for your PR.

Good catch for adding the `tabindex="0"` which makes possible to navigate into the Warning with the keyboard.

Unfortunately, testing with Safari and VoiceOver the Warning isn't announced. I guess because the editor sets focus on the newly inserted block. See in the screenshot below: focus is on the doubled 'More' block:

![Screenshot 2024-12-17 at 09 07 55](https://github.com/user-attachments/assets/b2c0d1c8-dcdc-4535-86a3-9616ece8406e)

I think a potential way to fix this would be setting focus on the Warning when it renders by using `useEffect` and `useRef`. 

I'll leave a few comments in the review as well. 
Worth noting there's a merge conflict because it appears this branch doesn't incorporate recent changes to the Warning component introduced in https://github.com/WordPress/gutenberg/pull/67675

afercia commented on packages/block-editor/src/components/warning/index.js:
```diff
@@ -14,14 +14,22 @@ import { moreVertical } from '@wordpress/icons';
+			<div
+				className={ clsx( className, 'block-editor-warning' ) }
+				role="alert"
+				aria-live="assertive"
```
role alert inherently works as an assertive live region so that `aria-live="assertive"` isn't necessary. `aria-live` doesn;t work well for elements that appear in the DOM anyways.

afercia commented on packages/block-editor/src/components/warning/index.js:
```diff
@@ -14,14 +14,22 @@ import { moreVertical } from '@wordpress/icons';
-						<div className="block-editor-warning__actions">
+						<div
+							className="block-editor-warning__actions"
+							aria-describedby="block-editor-warning-message"
```
Not sure this `aria-describedby` is necessary. It doesn't point to any ID anyways, please remove it.

afercia reviewed (changes requested):
Left some comments

sarthaknagoshe2002 committed:
Fix: resolve merge conflicts & ensure working in safari

sarthaknagoshe2002 commented:
@afercia I have now fixed things to make it work in safari.
Also, I have implemented the changes suggested by you.

| Chrome | Safari |
|-|-|
| <img width="1470" alt="Screenshot 2024-12-17 at 11 04 03 PM" src="https://github.com/user-attachments/assets/475b63c7-cfe8-424e-bc39-0d2be82404df" /> | <img width="1470" alt="Screenshot 2024-12-17 at 11 42 21 PM" src="https://github.com/user-attachments/assets/ba6dd759-cf72-49ef-91ef-49d27275980d" /> |

afercia commented on packages/block-editor/src/components/warning/index.js:
```diff
@@ -9,13 +9,28 @@ import clsx from 'clsx';
+		if ( null !== alertRef.current ) {
+			alertRef.current.focus();
+		}
+	}, [] );
```
Question: can this be simplified? I'm not sure there's the need to set the initial ref to `null` and then checking for it. To my understanding the optional chaining operator `?.` would make this much more concise. See an example here:

https://github.com/WordPress/gutenberg/blob/e4649fe6896123231de35da0875e93ac2920c23a/packages/block-library/src/video/tracks-editor.js#L182-L186

afercia commented on packages/upload-media/tsconfig.json:
```diff
@@ -8,13 +8,13 @@
 	},
 	"include": [ "src/**/*" ],
 	"references": [
-	  { "path": "../api-fetch" },
```
changes to this file should be reverted, probably a result of merge conflict or something

sarthaknagoshe2002 replied:
The pre-commit Git hook is enforcing linting even after reverting changes.  Are there alternative solutions?

afercia replied:
I submitted a quick PR https://github.com/WordPress/gutenberg/pull/68083 to fix it separately.
Once that is merged, please rebase this PR.

afercia commented on packages/block-editor/src/components/warning/index.js:
```diff
@@ -9,13 +9,28 @@ import clsx from 'clsx';
-					<p className="block-editor-warning__message">
+					<p
+						className="block-editor-warning__message"
+						ref={ alertRef }
```
I think the reg should be set on the container with tabIndex="0", otherwise moving focus will fail?

afercia commented:
> let me know if we should keep those changes too.

@sarthaknagoshe2002 yes I think those changes should be reverted.
Thanks for working on this, I left some new comments.

As a last step, when everything is done, I guess the snapshot of the failing test should be updated.

```
npm run test:unit -- --updateSnapshot --testPathPattern packages/block-editor/src/components/warning/test/index.js
```

afercia commented:
It's worth noting that _all_ Warnings will receive focus when they render, for example the 'invalid content' Warning and other ones. There may be edge cases but I think it's OK as default behavior.

![Screenshot 2024-12-18 at 09 19 18](https://github.com/user-attachments/assets/d6cc7a82-ed78-4436-ba5d-00ca62b0bfeb)

sarthaknagoshe2002 committed:
Merge branch 'WordPress:trunk' into fix/issue-62737

sarthaknagoshe2002 committed:
Merge branch 'WordPress:trunk' into fix/issue-62737

sarthaknagoshe2002 committed:
Fix: update test snapshot & optimize the useRef logic

sarthaknagoshe2002 committed:
Fix: use optional chaining

sarthaknagoshe2002 commented:
@afercia I believe this is now ready to be merged.

afercia reviewed (approved):
Looks good to me, thanks @sarthaknagoshe2002 !

afercia commented:
@WordPress/gutenberg-core as mentioned earlier, with this change _all_ Warnings will receive focus when they render. There may be edge cases but I think it's OK as default behavior and I'm going to merge this PR. Just wanted to be sure everyone is aware of this change in case there are edge cases to address in follow-ups.

afercia commented:
@t-hamano thanks for your feedback. To me, moving focus is the most sensible default, especially when the Warning contains action buttons. Keyboard users will benefit from being able to immediately access the actions.

> causing the current navigation focus to be lost

Yes I see your point. Should Warnings be rendered in the Patterns preview card in the first place? I didn't know of this behavior but to me it looks pretty weird. That is a preview, it should not contain interactive elements and they're so small anyways that I'd argue they're not usable at all. Screenshot:

![Screenshot 2024-12-19 at 10 31 50](https://github.com/user-attachments/assets/e81b1dc0-2134-4e56-b626-d4ded3c4c32f)

> when I open a post with multiple broken blocks, the last Warning component gets focus immediately.

Ideally, there should be only one Warning at a time. If the editor shows _multiple_ warnings at the same time, they should also be _all announced_ to screen reader users to provide an equivalent information. However, implementing a way to announce _all_ warnings at the same time could be potentially very noisy and confusing. I'm really not sure how to address this potential edge case.

afercia commented:
I'll create a separate issue for the Warnings in the preview cards.

Mamaduka commented:
I agree with @t-hamano that the `Warning` component shouldn't autofocus because of rendering issues. We should revert it.

## Why?

* The "focus because of render" is a tricky pattern in React; it's hard to get right, and because of that, it should be avoided.
* Multiple blocks can render a `Warning` component since they block a specific. The last one to render will steal the focus.
* These are caused by invalid or broken content when a post is rendered, not by user action. So they can be handled later.

TL;DR, let's avoid making similar changes to components that can be rendered multiple times and are widely used by block editors.

getdave commented:
I would also be in favour of reverting this. Autofocus on render is very likely to cause a number of knock on problems.

The other changes to make the warnings perceivable and more prominent to assistive tech seem like good improvements however.

afercia commented:
I'm not opposed to a revertl. However:

> The other changes to make the warnings perceivable and more prominent to assistive tech seem like good improvements however.

- The other improvements don't work without the focus being set on the Warning.
- The Warning may provide (and in most of the cases it does) action buttons. It's pointless and non-accessible rendering action buttons if keyboard users have to tab dozens and dozens of times to navigate to them.
- The Warning is a sort of 'alert' message. To make it usable for everyone it should be semantically an alert and should be announced by assistive tech and be easily navigable to with keyboard. Ideally, as with all alerts, there should be only one Warning rendered at a time.
- For accessibility, it's not uncommon to move focus to alerts and warning messages. It's a pretty common pattern. rather then removing this behavior we should find a way to make it solid and reliable.

Overall, the _usage_ of the Warning is highly problematic and all these points should be addressed. If anyone can think at better ways to address these points, that would be very welcome.

Mamaduka commented:
You can say that these warnings are mostly part of the content/canvas. Let's say you opened a documentation page with multiple warnings for experimental/unstable APIs. What are the expectations here? Do each of these warning messages try and move focus on itself, or are they just read as you get closer to actual experimental/unstable?

> rather then removing this behavior we should find a way to make it solid and reliable.

Ideally, we would have caught this broken behavior before merging this PR, but we are humans and sometimes miss things.

I'll include a full revert in my PR (#68133). Let's reopen the issue and continue working on a solution.

afercia commented:
Was thinking at a potential different approach, I will comment on the issue and reopen it.

Mamaduka commented:
Thanks, @afercia!

Meanwhile, my revert PR is ready to be merged - https://github.com/WordPress/gutenberg/pull/68133.

sarthaknagoshe2002 commented:
Since the focus was causing issues, we could have simply removed that part, as the warning was being announced by the screen reader in other browsers (except Safari) without the focus. This way, the rest of the functionality remains untouched.
